lib/grid.h

```diff
@@ -10,178 +10,941 @@
 #define DSL_GRID_H
 
 #include "cell.h"
+#include "utils.h"
 #include <Eigen/Dense>
+#include <vector>
+#include <memory>
+#include <fstream>
+
+#include "cereal/archives/binary.hpp" //serialization support
+#include "cereal/types/vector.hpp" //type support
+#include "cereal/types/eigen.hpp" //type support
 
 namespace dsl {
 
 struct EmptyData {};
+using std::shared_ptr;
+using std::vector;
+using namespace Eigen;
+
+
+//To check if a template parameter is shared_ptr
+//  usage: if(has_template_type<T, std::shared_ptr>::value)
+//           cout<<"is shared_ptr"<<endl;
+template < typename T,template <typename...> class Templated >
+struct has_template_type : std::false_type {};
+
+template < template <typename...> class T, typename... Ts>
+struct has_template_type<T<Ts...>, T> : std::true_type {};
+
 
 /**
  * An n-dimenensional grid consisting of abstract "cells", or elements
  * identified by a set of coordinates of type PointType, each cell
- * containing data of type DataType.
- * A grid provides instant access to the elements by maintaining
- * an n-dimensional array of pointers to cells. Cells that are empty,
- * e.g. that are inside obstacles, correspond to null pointers.
+ * containing data of type CellContent. The CellContent can directly be
+ * the intended data or shared_ptr to intended data. Using the shared_ptr
+ * is useful to avoid memory allocation until it's actually needed.
+ * CellContent=bool/double can be used to make a grid of occupancy or
+ * traversibility of a cell.
+ *
+ * Explanation of certain important variables:
+ *   id:   id for direct lookup in an array(or std::vector)
+ *   i:    dimension number(for grid over xyz coordinates i=0 refers to the x dimension)
+ *   gidx: grid index to locate a cell in a grid. For a 3d grid, gidx=(0,0,0) is the first cell along all dimension.
+ *   idx:  grid index along a particular dimension
  *
  * Note that this data structure is only viable up to a few dimensions,
  * e.g. dim=5 or 6.
+ *
+ * TODO: 1. Make sure origin lies at center of a cell
+ *       2. Check for bound in GetSlice and GetStack methods
+ *       3. Implement GetStack method that can perform occupancy dilation
+ *       4. Add clone methods to deep copy data
+ *
  */
-template < class PointType, class DataType = EmptyData>
- struct Grid {
+template < class PointType, class CellContent>
+class GridCore {
```

Comment:
There were two different classes Grid and Map, instantiated differently e.g. Grid<Vector3d, Matrix3d> and Map<3,bool> . While the cells in Grid stored Cell<Vector3d, Matrix3d>* , the cells in Map stored bools. 

To retain the same structure, I create this GridCore and aliased Grid and Map from it. Please advice how should  I structure them both finally.